lxd/instance/drivers/driver_qemu.go

```diff
@@ -2503,36 +2453,66 @@ func (d *qemu) deviceAttachNIC(deviceName string, netIF []deviceConfig.RunConfig
 	return nil
 }
 
-// getPCIHotplug returns an available PCI device ID that can be used for hotplugging.
-func (d *qemu) getPCIHotplug() (string, error) {
+// getPCIHotplug returns an available PCI device bus name and address that can be used for hotplugging.
+// multi will always be false but is included for compatibility with busAllocator interface.
+func (d *qemu) getPCIHotplug() (busName string, busAddress string, multi bool, err error) {
 	// Check if the agent is running.
 	monitor, err := qmp.Connect(d.monitorPath(), qemuSerialChardevName, d.getMonitorEventHandler())
 	if err != nil {
-		return "", err
+		return "", "", false, err
 	}
 
-	// Get the current PCI devices.
-	devices, err := monitor.QueryPCI()
-	if err != nil {
-		return "", err
+	hasUsedBridges := func(devices []qmp.PCIDevice) bool {
+		for _, dev := range devices {
+			if len(dev.Bridge.Devices) > 0 {
+				return true
+			}
+		}
+
+		return false
 	}
 
-	for _, dev := range devices {
-		// Skip built-in devices.
-		if dev.DevID == "" {
-			continue
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	for {
+		// Get the current PCI devices.
+		devices, err := monitor.QueryPCI()
+		if err != nil {
+			return "", "", false, fmt.Errorf("Failed to query PCI devices: %w", err)
 		}
 
-		// Skip used bridges.
-		if len(dev.Bridge.Devices) > 0 {
+		if !hasUsedBridges(devices) {
+			// Wait for PCI bridges to be created (with the internal devices LXD adds) to ensure QEMU
+			// has had time to initialise the QMP query-pci output.
+			// Otherwise we'll end up trying to hotplug into a used PCI slot.
+			d.logger.Debug("Waiting for PCI to initialize")
+
+			select {
+			case <-ctx.Done():
+				return "", "", false, fmt.Errorf("Timed out waiting for PCI to initialize: %w", ctx.Err())
+			case <-time.After(500 * time.Millisecond):
```

Comment:
If the first iteration barely misses it, it adds a long delay to the boot. Would it be OK to burn a bit more CPU to more actively poll here and hopefully break out earlier?

Reply:
https://github.com/canonical/lxd/pull/16067/